HUB-218: Read List of Services From Hub

chrisclayson opened the pull request:
Add New Service to read list of services from Hub that can be injected into controllers.

thomasalanlee commented on stub-idp/src/integration-test/java/uk/gov/ida/apprule/ServiceListServiceIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class ServiceListServiceIntegrationTest extends IntegrationTestHelper {
```
This is probably a bit confusing but our integration tests usually test the app end to end (rather than testing a single service). Perhaps this test could be ported to a unit test?

thomasalanlee commented on stub-idp/src/integration-test/java/uk/gov/ida/apprule/ServiceListServiceIntegrationTest.java:
```diff
@@ -0,0 +1,98 @@
+        Optional<Service[]> services = listService.getServices();
+
+        assertTrue(services.isPresent());
+        assertEquals(services.get().length, 3);
```
We should use the AssertJ matchers e.g. assertThat to keep the test consistent with the other tests

thomasalanlee commented on stub-idp/src/main/java/uk/gov/ida/stub/idp/services/ServiceListService.java:
```diff
@@ -0,0 +1,45 @@
+        return readListFromHub();
+    }
+
+    private Optional<Service[]> readListFromHub(){
```
This should probably not return an Optional. Its more conventional to return something from the Collections lib and handle the fact that it can be of` .size` 0 in the calling class

chrisclayson commented:
Removed the 'Optional' return type on service and also removed 'integration' test altogether as the unit test did much the same thing using a different mock.

chrisclayson commented:
Now returning List<Service> rather than array.

chrisclayson commented on stub-idp/src/main/java/uk/gov/ida/stub/idp/services/ServiceListService.java:
```diff
@@ -0,0 +1,48 @@
+
+    private List<Service> readListFromHub(){
+        try {
+            List<Service> services = Arrays.asList(jsonClient.get(singleIdpConfiguration.getServiceListUrl(), Service[].class));
```
I had already tried that, resulted in a `null` being returned in the tests ... I now realise, it's cos I didn't refactor the mock to match.

chrisclayson commented:
OK ... amended!

thomasalanlee commented on stub-idp/src/test/java/uk/gov/ida/stub/idp/services/ServiceListServiceTest.java:
```diff
@@ -0,0 +1,92 @@
+    @Test
+    public void getServiceListWhenFeatureEnabled_ReturnsAListOfServices() throws FeatureNotEnabledException {
+        when(singleIdpConfiguration.isEnabled()).thenReturn(true);
+        when(jsonClient.get(eq(uri), Matchers.<GenericType<List<Service>>>any())).thenReturn(Arrays.asList(new Service[] { service1, service2, service3, service4, service5}));
```
You should be able to replace
Arrays.asList(new Service[] { service1, service2, service3, service4, service5}));
With
Arrays.asList(service1, service2, service3, service4, service5);
